ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Mempool.hs

```diff
@@ -128,6 +128,10 @@ instance ShelleyBasedEra era
 
   txForgetValidated (ShelleyValidatedTx txid tx) = ShelleyTx txid (SL.extractTx tx)
 
```

Comment:
Here we do have a problem and I would love to discuss it quickly.

Instance for `LedgerSupportsMempool` was written for all blocks which are "shelley based". However information about execution units is only held in Alonzo types (e.g fields like `totExunits`). We can not add constraints to `ShelleyBasedEra` like `HasField "totExunits"`.
It seems that we have to just create a dedicated instance for `LedgerSupportsMempool` that is crafted just for Alonzo, and potentially more general one for the rest of the shelley eras. Or maybe there is some other, smarter way to do things here

Reply:
I haven't worked through it, but my suggestion would be to add the new methods to the `ShelleyBasedEra` class instead of to the `LedgerSupportsMempool` class. Two benefits in light of this current diff:

  1) We don't need to define these methods for Byron, ByronDual, HardForkBlock, etc.
  
  2) The `ShelleyBasedEra (AlonzoEra c)` instance can have the `HasField "totExunits"` constraint and the other eras can just use `mempty`, with no constraint.
  
Give that a try and let me know what you think.